### PR TITLE
Store Greaselion scripts by sender id and tab id

### PR DESCRIPTION
**Pull request by emerick**
**Note**: Don't merge until https://github.com/brave/brave-core/pull/6841 is merged, as this is a backwards-incompatible change.

Resolves https://github.com/brave/brave-browser/issues/12090

Related Greaselion PR: https://github.com/brave/brave-site-specific-scripts/pull/22

To support splitting Greaselion scripts into "base" and "tipping" components (to allow panel-based tipping even when inline tipping is disabled), we needed to modify how we map Greaselion ports. Previously, we assumed one Greaselion script per page, so the mapping was `port` -> `tabId`. Since there can now be more than one Greaselion script per page, the mapping is `port` -> `tabId:senderId` (where `senderId` is the Greaselion script's extension ID).

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Requested a security/privacy review as needed.
- [x] Added appropriate QA labels (`QA/Yes` or `QA/No`) to the associated issue
- [x] Added appropriate release note labels (`release-notes/include` or `release-notes/exclude`) to the associated issue
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Clean profile
- Enable Rewards
- Enable inline tipping for all media publishers
- Visit https://github.com/brave/brave-core/commits
- Ensure that inline tip button shows up and works as expected
- Ensure that you can tip via the Rewards panel
- Disable inline tipping for Github
- Refresh Github page
- Ensure that inline tip button no longer appears
- Ensure that you can still tip via the Rewards panel

Other media publishers that support inline tips (Twitter and Reddit) can be tested in a similar way.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts [hunk 1]
```diff
@@ -341,16 +358,16 @@ chrome.runtime.onConnectExternal.addListener((port: chrome.runtime.Port) => {
         }
         case 'RegisterOnCompletedWebRequest': {
           const data = msg.data as RegisterOnCompletedWebRequest
-          handleRegisterOnCompletedWebRequest(tabId, msg.mediaType, data)
+          handleRegisterOnCompletedWebRequest(tabId, senderId, msg.mediaType, data)
```
Comment (petemill) [inline components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts:361]: Haven't read through this whole module so maybe I'm missing something about how the the event handlers interact with the Maps, but is there a reason we're not just passing the `port` here and avoid the concat in each of these functions? Just a nit I suppose.

Comment (emerick) [reply components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts:361]: Yep good point, will fix.

**Comment by zenparsing** [comment]
Couple of thoughts:

- It's possible to store object references as the keys in a `Map`. Would it work or be preferable to use a `Map<Port, T>` instead of `Map<string, T>`?
- Except for `portsByTabIdSenderId`, we aren't calling delete on these maps. Could this cause a slow memory leak? Do we need to clean up these other map entries on disconnect as well?
- If you did end up using `Map<Port, T>`, then you could also consider using `WeakMap<Port, T>` which will take care of releasing the map entry when the object reference key is no longer reachable.

**Comment by emerick** [comment]
@zenparsing I updated the code to call delete the relevant entries from all of the maps on disconnect, great catch. I think I'll continue with the string key for now, as then all of the maps are keyed the same.

**Review by NejcZdovc — APPROVED** [review]

**Review by zenparsing — APPROVED** [review]

**Comment by emerick** [comment]
CI passed, test-install failed on Mac and a couple of currently expected test failures on Windows.

**Comment by emerick** [comment]
This functionality will be available in Greaselion `1.0.34`.

**Comment by LaurenWags** [comment]
Verified using

```
Brave	1.18.18 Chromium: 87.0.4280.27 (Official Build) nightly (x86_64)
Revision	d8c1fe98de8d9c6bd46ebe8b0cc9d2b6bdccca1e-refs/branch-heads/4280@{#556}
OS	macOS Version 10.14.6 (Build 18G6032)
```

Per https://github.com/brave/brave-core/pull/6889#issuecomment-718022208, confirmed I had correct component version:
Screen Shot 2020-10-29 at 4 02 08 PM

Verified STR (both cases) from https://github.com/brave/brave-browser/issues/12090#issue-719427683

Verified test plan from https://github.com/brave/brave-core/pull/6889 for Twitter

Verified test plan from https://github.com/brave/brave-core/pull/6889 for GitHub

Verified test plan from https://github.com/brave/brave-core/pull/6889 for Reddit

Confirmed if content-level tips are unchecked for Twitter, in-line tip buttons still generally display on Reddit and GitHub. Tried other combinations as well ex. Twitter/Reddit off, GH on, etc. Confirmed able to tip from panel when content-level tips were disabled.

Confirmed in-line tip button will appear again once toggled on (need page refresh if site is already open) for Twitter, Reddit, GitHub content-level tipping. Confirmed able to tip from in-line tip button once it was re-enabled.

Examples where content-level tips are unchecked on brave://rewards and panel tipping is still available:
twitter
reddit
gh

**Pull request merged**